### DISCUSSION TITLE
Incorporating Lithium Plating effects for DFN model using 'Ai2020' parameter set, and some other questions.

### DISCUSSION THREAD

**Discussion by HassanSewailem** [pybamm-team/PyBaMM#2340] [Q&A] [answered]
Hello, 

I have a few questions. Hope you can solve them! 

Question 1:
I seem to get an error when I set the model options for lithium plating to anything other than 'none' when using P2D model with the parameter set 'Ai2020'. 
I get the error,
KeyError: "'Exchange-current density for plating [A.m-2]' not found. Best matches are ['SEI reaction exchange current density [A.m-2]', 'Negative electrode exchange-current density [A.m-2]', 'Positive electrode exchange-current density [A.m-2]']"

How could I incorporate lithium plating effects while using this specific parameter set? The code can be found at the bottom of this message. 

Question 2: 
I seem to be having troubles using the "ParameterValues()" function for any parameter set. For example when I use,
 params = pybamm.ParameterValues("Ai2020"), it gives the error: parameter Ai2020 not found. 

I always use the other method being,
chemistry = pybamm.parameter_sets.Ai2020
params = pybamm.ParameterValues(chemistry = chemistry)

However, with this method, some parameter sets are unavailable, for example OKane2022.  Thus, I would like to know why the first method is not working for me. 

Question 3:
I am curious to know which papers were used to calculate the capacity lost when doing cycle testing, for the SPMe and the DFN models. 

Question 4: 
I see the SEI parameters do not change, at least for the papers I have checked ('Chen2020', 'Ai2020', 'Ecker2015'). Would that cause inaccuracies in the diffusion dynamics? Since the battery chemistries are different.  Also, which paper are these parameters originally from? Since I am currently trying to obtain the correct lithium plating and SEI parameters to be applicable on the battery used in 'Ai2020', in order to implement my own degradation model from the paper [https://www.semanticscholar.org/paper/Modeling-of-lithium-plating-induced-aging-of-from-Yang-Leng/ebe51c8e4f7356737cca0798b08af55c9d598c12](url). If such parameters for equations involving SEI and Lpl concentrations, as well as film thickness are available, that would be great.

The code for the first question can be found below. 
 Thank you in advance for your time!  

---------------------------------------------------------------------------------------------------------------------------------------------------
import pybamm 
import numpy as np 
import matplotlib.pyplot as plt

## Model and parameter set initiation 
options = {"SEI": "ec reaction limited",
          "SEI film resistance": "distributed",
          "lithium plating":"irreversible"}
model = pybamm.lithium_ion.DFN(options=options) ## P2D model
model.options.possible_options
model.options.print_options() ## Print possible options in model 
#params = pybamm.ParameterValues("Ai2020")
chemistry = pybamm.parameter_sets.Ai2020 ## Use Ai2020 parameter set,  
params = pybamm.ParameterValues(chemistry = chemistry)
#params.search("width")
## Depleting the battery first and setting initial conditions to the ones 
discharge_exp = pybamm.Experiment(["Discharge at C/10 until 3V"]) ## Setting up experiment to deplete battery at low current rate 
discharge_sim = pybamm.Simulation(model, parameter_values=params, experiment=discharge_exp) ## Simulating experiment
discharge_sol = discharge_sim.solve(calc_esoh=False) 

# MODEL RE-INITIALIZATION: #############################################################
# Now initialize the model with the solution of the discharge, and then charge 
# We could also do this inplace by setting inplace to True, which modifies the original
# model in place
new_model = model.set_initial_conditions_from(discharge_sol, inplace=False)

########################################################################################

**Comment by valentinsulzer** [comment, marked as answer]
1: you have to manually add the lithium plating parameters for that parameter set, e.g. using the same values as the other parameter sets (or better values if you have them)
2: try updating to the latest version of PyBaMM
3: The summary variable "Capacity [A.h]" is calculated as a function of total cyclable lithium and electrode capacities (related to LLI/LAM). See this example for details and reference: https://github.com/pybamm-team/PyBaMM/blob/develop/examples/notebooks/models/electrode-state-of-health.ipynb. This capacity ignores C-rate and effect of internal resistance.
The summary variable "Measured capacity [A.h]" is simply coulomb counting between the voltage limits (i.e. returns different values at different C-rates). https://github.com/brosaplanella/SPMe_SR is quite nice for showing some of the ways to plot this
4: any degradation parameters should be taken with a grain of salt. At the moment they are representative only, *not* fitted for that particular dataset

**Reply by HassanSewailem** [reply]
Thank you for taking the time to reply.

For the first question, I m currently try to add the lithium plating parameters from OKane2022 to the Ai2020 parameter set. While I did manage to add all the constant parameters, I cannot seem to find the proper syntax to add a function parameter to a parameter set. for example the plating exchange current density function. The code can be found below,

import pybamm
import numpy as np
import matplotlib.pyplot as plt

options = {"SEI": "ec reaction limited",
          "SEI film resistance": "distributed",
          "lithium plating":"none"}

model = pybamm.lithium_ion.DFN(options=options) ## P2D model
model.options.possible_options
model.options.print_options() ## Print possible options in model
params = pybamm.ParameterValues("Ai2020")

discharge_exp = pybamm.Experiment(["Discharge at C/10 until 3V"]) ## Setting up experiment to deplete battery at low current rate
discharge_sim = pybamm.Simulation(model, parameter_values=params, experiment=discharge_exp) ## Simulating experiment
discharge_sol = discharge_sim.solve(calc_esoh=False)

new_model = model.set_initial_conditions_from(discharge_sol, inplace=False)

params.update({'Initial plated lithium concentration [mol.m-3]': 0.0,
'Exchange-current density for plating [A.m-2]': pybamm.FunctionParameter(plating_exchange_current_density_OKane2020), #error
'Lithium plating kinetic rate constant [m.s-1]': 1e-9,
'Lithium plating transfer coefficient': 0.65,
'Typical plated lithium concentration [mol.m-3]' : 1000.0}, check_already_exists=False)

params.search("plat")
params.values()

**Reply by valentinsulzer** [reply]
You should pass in the function object itself, e.g.
```python3
def plating_function(c_e, c_Li, T):
    k_plating = 0.001
    return pybamm.constants.F * k_plating * c_e

params.update({'Exchange-current density for plating [A.m-2]': plating_function})
```

**Reply by HassanSewailem** [reply]
This worked, thank you.
